### PR TITLE
Support diamond inheritance of sealed interfaces in SealedClassSerializer

### LINKED ISSUE #1937: Multiple inheritance bug
**Describe the bug**

There is an issue with multiple inheritance, using interfaces, when multiple paths leads to the same serializable class.

**To Reproduce**
```kotlin
import kotlinx.serialization.Serializable
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json

fun main(args: Array<String>) {
    Json.encodeToString(Z(X()))
}

@Serializable
sealed interface A {}

@Serializable
sealed interface B: A {}

@Serializable
sealed interface C: A {}

@Serializable
class X: B, C


@Serializable
class Y(val b: B)

@Serializable
class Z(val a: A)
```

**Expected behavior**

Expected that the serialization just work, but it crashes with
`Element with name 'X' is already registered`

**Environment**
 - Kotlin version: 1.6.21
 - Library version: 1.6.21
 - Kotlin platforms: JVM, JS (BOTH)

### PR DESCRIPTION
**Pull request by sandwwraith**
Fixes #1937

### CODE REVIEW

File: core/commonTest/src/kotlinx/serialization/SealedDiamondTest.kt [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+import kotlinx.serialization.descriptors.*
+import kotlin.test.*
+
+class SealedDiamondTest {
+
+    @Serializable
+    sealed interface A {}
+
+    @Serializable
+    sealed interface B: A {}
```
Comment (qwwdfsad) [inline core/commonTest/src/kotlinx/serialization/SealedDiamondTest.kt:12]: Our code-style implies whitespace prior to `:`

File: core/commonTest/src/kotlinx/serialization/SealedDiamondTest.kt [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+    sealed interface C: A {}
+
+    @Serializable
+    @SerialName("X")
+    class X: B, C
+
+    @Test
+    fun testMultipleSuperSealedInterfaces() {
+        val subclasses = A.serializer().descriptor.getElementDescriptor(1).elementDescriptors.map { it.serialName }
+        assertEquals(listOf("X"), subclasses)
```
Comment (qwwdfsad) [inline core/commonTest/src/kotlinx/serialization/SealedDiamondTest.kt:24]: Could you please also add a test that ensures that every type in the diamond is properly serialized and deserialized via `A`, `B` and `C` types?

File: core/commonTest/src/kotlinx/serialization/SealedDiamondTest.kt (hunk 1 above)
Comment (sandwwraith) [reply core/commonTest/src/kotlinx/serialization/SealedDiamondTest.kt:12]: Apparently reformatter doesn't do that for some reason

File: formats/json/commonTest/src/kotlinx/serialization/features/sealed/SealedDiamondTest.kt [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+    @Test
+    fun testMultipleSuperSealedInterfacesDescriptor() {
+        val subclasses = A.serializer().descriptor.getElementDescriptor(1).elementDescriptors.map { it.serialName }
+        assertEquals(listOf("X"), subclasses)
+    }
+
+    @Test
+    fun testMultipleSuperSealedInterfaces() {
+        @Serializable
+        data class Carrier(val a: A, val b: B, val c: C)
```
Comment (qwwdfsad) [inline formats/json/commonTest/src/kotlinx/serialization/features/sealed/SealedDiamondTest.kt:32]: 👍

**Review by qwwdfsad — APPROVED** [review]

**Comment by StylianosGakis** [comment]
@mgroth0 I guess this https://github.com/Kotlin/kotlinx.serialization/releases/tag/v1.4.0-RC is what you were looking for

**Pull request merged**